perma_web/fabfile/dev.py

```diff
@@ -573,3 +573,113 @@ def check_s3_hashes():
+        for key in storages:
+            storage = storages[key]['storage']
+            with open('/tmp/perma_storage_cache_{0}.txt'.format(key), 'w') as tmp_file:
+                if hasattr(storage, 'bucket'):
+                    # S3
+                    for f in tqdm(storage.bucket.list('generated/warcs/')):
+                        # etag is a string like u'"3ea8c903d9991d466ec437d1789379a6"', so we need to
+                        # knock off the extra quotation marks
+                        # note that etags are not always md5sums, but should be in these cases; we can rewrite
+                        # or replace md5hash if necessary
+                        path = f.key[(len(storage.location)):]
+                        hash = f.etag[1:-1]
+                        tmp_file.write("{0}\t{1}\n".format(path, hash))
+                        storages[key]['lookup'][path] = hash
+                else:
+                    if hasattr(storage, '_root_path'):
+                        # SFTP
+                        base = storage._root_path
+                    else:
+                        # local file storage -- are there other possibilities to consider?
+                        base = storage.base_location
+                    for f in tqdm(storage.walk(os.path.join(base, 'warcs'))):
+                        # os.walk: "For each directory in the tree rooted at directory top (including top itself),
+                        # it yields a 3-tuple (dirpath, dirnames, filenames)" -- there should be exactly one WARC
+                        # per directory, so:
+                        if len(f[2]) == 1:
+                            full_path = os.path.join(f[0], f[2][0])
+                            path = full_path[len(base):]
+                            hash = md5hash(full_path, storage)
+                            tmp_file.write("{0}\t{1}\n".format(path, hash))
+                            storages[key]['lookup'][path] = hash
+    else:
+        print("Reading storage caches ...")
+        for key in storages:
+            with open('/tmp/perma_storage_cache_{0}.txt'.format(key)) as f:
+                for line in f:
+                    path, hash = line[:-1].split("\t")
+                    storages[key]['lookup'][path] = hash
+
+    # now check ground truth against storage lookup tables
+    print("Comparing link cache against storage caches ...")
+    with open(link_cache) as f:
+        for line in f:
+            path = line[:-1]
+            file_present = True
+            for key in storages:
+                if path not in storages[key]['lookup']:
+                    print("{0} not in {1}".format(path, key))
+                    file_present = False
+            if file_present and len(storages) > 1:
+                hashes = []
+                for key in storages:
+                    hashes.append(storages[key]['lookup'][path])
+                if hashes.count(hashes[0]) != len(hashes):
```

Comment:
Wow! That's impressively faster. Cool. Maybe the comment would be good, especially since we are talking about a list of length 2 rather than long lists like they were comparing?